pkg/websocket/connection.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package websocket
+
+import (
+	"bytes"
+	"encoding/gob"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/gorilla/websocket"
+)
+
+var (
+	// ErrConnectionNotEstablished is returned by methods that need a connection
+	// but no connection is already created.
+	ErrConnectionNotEstablished = errors.New("connection has not yet been established")
+)
+
+// RawConnection is an interface defining the methods needed
+// from a websocket connection
+type RawConnection interface {
+	WriteMessage(messageType int, data []byte) error
+	NextReader() (int, io.Reader, error)
+	Close() error
+}
+
+// ManagedConnection represents a websocket connection.
+type ManagedConnection struct {
+	connFactory    func() (RawConnection, error)
```

Comment:
Often for testing I'll simplify this by having:
```golang
var connFactory = func() { ... }
```

Then the test code can set the *private* global and the testability doesn't leak into the API.